### PR TITLE
`Exam mode`: Handle quiz pool in exam conduction

### PR DESCRIPTION
**Pull request by rriyaldhi**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Currently, if an instructor wants every student to have a different set of quiz questions for each student, the instructor must create one quiz exercise for one set of quiz questions. The higher the number of sets of quiz questions, the more quiz exercises must be prepared by the instructor. This is cumbersome if the instructor wants to have a high variance of quiz questions for the students.

To improve the user experience for instructors for such requirements, the quiz pool configuration is created ([#6402](https://github.com/ls1intum/Artemis/pull/6402)). Quiz pool configuration enables the instructor to put quiz questions into a group. The example configuration is shown below:

image

Then, from each group, one question is randomly assigned to each student ([#7583](https://github.com/ls1intum/Artemis/pull/7583)). During the exam, the quiz questions will be displayed on a separate page **(Quiz)**.

Hence, the exam participation page needs to be modified to display the quiz questions in the **Quiz** section.

### Description

**Server**

- Add quizExamMaxPoints and randomizeQuizExamQuestionsOrder properties to the Exam model. The quizExamMaxPoints is used by the client to determine if an exam contains a **Quiz** section or not. If there is a **Quiz**  section, the randomizeQuizExamQuestionsOrder property is used to determine if the quiz questions are randomized or not.

- Add fetchAllQuizQuestions and findWithEagerQuizQuestionsById method to StudentExamRepository to load quiz questions that are assigned to a specific StudentExam.

- Add findByExamId to QuizPoolService to be used by StudentExamResource to determine if there is a quiz pool configured for the given exam.

- Add getCourseByQuizQuestion to QuizPoolService to get the Course given QuizQuestion. The Course is used in FileResource.getDragAndDropBackgroundFile and FileResource.getDragItemFile methods.

- Update FileResource.getDragAndDropBackgroundFile and FileResource.getDragItemFile to call QuizPoolService.findByExamId if QuizQuestion is configured in QuizPool.

- Update QuizPoolResource.updateQuizPool to consume MULTIPART_FORM_DATA_VALUE to enable uploading files for DnD question.

- Call ExamQuizQuestionsGenerator.generateQuizQuestionsForExam in StudentExamService.generateTestRun to assign quiz questions during a test run.

- Call QuizPoolService.findByExamId during exam conduction to check if there is a quiz pool configured. If yes, call StudentExamRepository.fetchAllQuizQuestions to fetch quiz questions that are assigned to the given student exam and set quizExamMaxPoints and randomizeQuizExamQuestionsOrder to the exam object.

- Update DatabaseQueryCountTest.testExamQueryCount to add one more query call to check if there is a quiz pool configured for the given exam.

**Client**

- Add isQuizExamPage property to ExamPage model class. The property is used to determine if the student is currently on the **Quiz** section.

- Add quizExamMaxPoints randomizeQuizExamQuestionsOrder properties to the Exam model and the quizQuestions property to the StudentExam model. The quizExamMaxPoints is used to determine if there is a separate **Quiz** section for the exam. The quizQuestions property is passed to QuizExamExercise to be displayed to the students.

- Add quizExam model class. This model holds all information related to the **Quiz** section such as the questions and submission.

- Add quizQuestions property to the StudentExam model. The property holds the quizQuestions that are assigned to the Student.

- Update the parameter of the examNavigationBarComponent.changePage method call in StudentExamTimelineComponent to have 1 more false value for quizExam. The false value means the student does not change the page to the **Quiz** section

- Update ExamNavigationBarComponent to have a **Quiz** button if quizExam exists (quizPool is configured for the exam). The text at the top left of the page is also adjusted to be **Quiz** for the **Quiz** section.

- Update ExamParticipationComponent to have the **Quiz** section page.

- Add createQuizExam method in exam utils. The method is used to construct QuizExam object.

- Update ExamExerciseOverviewPage to also have a QuizExam row if quizExam exists.

- Add quizPoolId property to QuizQuestions. This property is used to get a Course given a QuizQuestion.

- Move handleDndQuizFileCreation, handleDndQuizDragItemsCreation, handleDndQuizFileUpdates, validateQuizExerciseFiles, saveDndQuestionBackground, and saveDndDragItemPicture methods from QuizExerciseService to QuizService, so the methods can be reused for QuizPoolService.

- Update QuizPoolComponent to pass files to QuizPoolService.update method.

- Update QuizPoolService.update to pass FormData which contains quizPool and files.

- Add translations of the Quiz Exam title and navigation title.

#### Exam Mode Testing

**Note: Detailed steps can be seen on the screencast in the screenshot section below**

Prerequisites:
- 1 Instructor (artemis_test_user_[11-15])
- 1 Exam with Quiz Pool:
  - **TS2**: [https://artemis-test2.artemis.cit.tum.de/course-management/22/exams/22](https://artemis-test2.artemis.cit.tum.de/course-management/22/exams/22)
  - **TS3**: [https://artemis-test3.artemis.cit.tum.de/course-management/25/exams/20](https://artemis-test3.artemis.cit.tum.de/course-management/25/exams/20)
  - **TS5**: [https://artemis-test5.artemis.cit.tum.de/course-management/15/exams/19](https://artemis-test5.artemis.cit.tum.de/course-management/15/exams/19)
  - **TS6**: [https://artemis-test6.artemis.cit.tum.de/course-management/24/exams/6](https://artemis-test6.artemis.cit.tum.de/course-management/24/exams/6)
- 1 Exam without Quiz Pool

1. Log in to Artemis as an instructor (artemis_test_user_[11-15]).
2. Go to the exam details page of specified exams above.
3. Reset the exam. (Could be failing if the exam is in progress, resetting the second time will work)
4. Edit the start and end of the exam so we can test the conduction
5. Generate student exams.
6. Log out and log in as a student (artemis_test_user_[6-10]).
7. Start the exam.
8. Verify that there is a **Quiz** section at the beginning of the exam and that the quiz questions are displayed correctly.
9. Verify that another exercise on the exam is also displayed correctly.
10. Repeat the steps 2-8 for another Exam without Quiz Pool

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| Exam.java | 92% | ✅                           |
| QuizQuestion.java | 97% | ✅                           |
| StudentExamRepository.java | 90% | ✅                           |
| QuizExerciseService.java | 100% | ✅                           |
| QuizPoolService.java | 97% | ✅                           |
| QuizService.java | 92% | ✅                           |
| ExamService.java | 92% | ✅                           |
| StudentExamService.java | 94% | ✅                           |
| ExamResource.java | 95% | ✅                           |
| StudentExamResource.java | 86% | ✅                           |
| FileResource.java | 71% | ✅                           |
| QuizExerciseResource.java | 93% | ✅                           |
| QuizPoolResource.java | 100% | ✅                           |
| exam-status.component.ts | 86.84% | ✅              |
| exam-checklist.component.ts | 95.91% | ✅              |
| exam-checklist.service.ts | 100% | ✅              |
| student-exam-timeline.component.ts | 92.62% | ✅              |
| exam-navigation-bar.component.ts | 88.39% | ✅              |
| exam-participation.component.ts | 89.75% | ✅              |
| exam-participation.service.ts | 87.78% | ✅              |
| exam.utils.ts | 100% | ✅              |
| exam-exercise-overview-page.component.ts | 90.24% | ✅              |
| quiz-pool.component.ts | 92.66% | ✅              |
| quiz-pool.service.ts | 92.85% | ✅              |

### Screenshots

image

https://github.com/ls1intum/Artemis/assets/10885503/bf0c027f-677d-447d-8ed0-39c7e631e7e0

## Summary by CodeRabbit

- **New Features**
	- Introduced the ability to randomize the order of quiz questions in exams.
	- Added support for handling file uploads and updates for quiz questions.
	- Enhanced quiz exams with maximum points calculation and better integration into various components.
- **Enhancements**
	- Improved quiz question generation for test runs.
	- Streamlined the handling of quiz questions in exams, including fetching and setting quiz exam properties.
- **Bug Fixes**
	- Adjusted validation logic for exercise groups in exams to include points calculation for quiz exams.
- **Refactor**
	- Updated several services and repositories to support new quiz functionalities and file handling.
- **Tests**
	- Added and modified tests to cover new functionalities and changes in quiz and exam handling.

### CODE REVIEW

**Review by vinceclifford — DISMISSED** [review]
Works as expected, tested during testing session

**Review by lennart-keller — DISMISSED** [review]
Tested in testing session, works as expected.

**Review by valentin-boehm — DISMISSED** [review]
Tested in Testing Sessions

**Comment by Ceiridge** [comment]
Tested on ts3.
The quiz questions do not appear at all in an exam test run. Is this intentional?

image
image

**Review by mateusmm01 — APPROVED** [review]
Tested in testing session and looks good

**Review by lennart-keller — COMMENTED** [review]
Code looking good overall. Just a few optional questions/suggestions :)

File: src/main/java/de/tum/in/www1/artemis/service/QuizPoolService.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizPoolService.java]: Updated to use Optional

File: src/main/webapp/app/entities/quiz-exam-exercise.ts
Comment (rriyaldhi) [reply src/main/webapp/app/entities/quiz-exam-exercise.ts:39]: The values are intended and displayed during the exam conduction. But now, I refactored it a bit to use translated values instead of hard-coded values.

**Comment by rriyaldhi** [comment]
> Tested on ts3. The quiz questions do not appear at all in an exam test run. Is this intentional?
> 
> image image

Thanks for catching this. This is fixed now:

image

**Review by Santia-go — DISMISSED** [review]
Generates the exam works accordingly. 

image

Resolve the exam from a student perspective also works. 

image


image

All worked as expected.

File: src/main/webapp/app/exam/participate/exercises/exercise-overview-page/exam-exercise-overview-page.component.html
Comment (milljoniaer) [inline src/main/webapp/app/exam/participate/exercises/exercise-overview-page/exam-exercise-overview-page.component.html:44]: Can we somehow avoid these function calls?

File: src/main/webapp/app/exam/participate/exam-navigation-bar/exam-navigation-bar.component.html
Comment (milljoniaer) [inline src/main/webapp/app/exam/participate/exam-navigation-bar/exam-navigation-bar.component.html:58]: is this function call avoidable?

**Review by milljoniaer — APPROVED** [review]
Code looks good, I just left a few comments regarding function calls in HTML.

File: src/main/webapp/app/exam/participate/exercises/exercise-overview-page/exam-exercise-overview-page.component.html
Comment (rriyaldhi) [reply src/main/webapp/app/exam/participate/exercises/exercise-overview-page/exam-exercise-overview-page.component.html:44]: Yes, it's avoidable. Thanks for catching this. The method call on the template is removed.

File: src/main/webapp/app/exam/participate/exam-navigation-bar/exam-navigation-bar.component.html
Comment (rriyaldhi) [reply src/main/webapp/app/exam/participate/exam-navigation-bar/exam-navigation-bar.component.html:58]: Yes, it's avoidable. Thanks for catching this. The method call on the template is removed.

**Review by milljoniaer — DISMISSED** [review]
Changes LGTM 🚀

**Review by basak-akan — COMMENTED** [review]
Tested during the testing session and we found a small issue;
After answering questions and then save&continue on quiz exercise, still the pen icon is shown with a exercise saved tooltip

image

**Comment by rriyaldhi** [comment]
> Tested during the testing session and we found a small issue; After answering questions and then save&continue on quiz exercise, still the pen icon is shown with a exercise saved tooltip
> 
> image

This is fixed on the next PR: [#7748](https://github.com/ls1intum/Artemis/pull/7748) since it is hardcoded anyway on this PR.

**Comment by rriyaldhi** [comment]
> Tested during the testing session and we found a small issue; After answering questions and then save&continue on quiz exercise, still the pen icon is shown with a exercise saved tooltip
> 
> image

This is fixed now:
image

**Review by basak-akan — DISMISSED** [review]
Thank you for fixing the small issue

**Review by MarkusPaulsen — APPROVED** [review]
Tested on TS3

**Review by Jan-Thurner — DISMISSED** [review]
Tested on TS3

**Review by coolchock — APPROVED** [review]

**Review by nata-81 — APPROVED** [review]
tested during test session, on ts3, works as expected.

**Review by konrad2002 — DISMISSED** [review]
manually tested on TS3, worked as expected

**Review by Jan-Thurner — DISMISSED** [review]
reapproval after mergeconflict

**Review by coolchock — DISMISSED** [review]
re-approve after merge conflict

**Review by milljoniaer — APPROVED** [review]
reapprove after code changes 🚀

**Review by Santia-go — DISMISSED** [review]
The merge conflict has been solved. 
Re-approved

**Pull request closed without merging**
